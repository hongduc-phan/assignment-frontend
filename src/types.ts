// Action types
export const ADD_PRODUCT = 'ADD_PRODUCT'
export const REMOVE_PRODUCT = 'REMOVE_PRODUCT'
export const TOGGLE_DIALOG = 'TOGGLE_DIALOG'

export const ADD_ITEM_TO_CART = 'ADD_ITEM_TO_CART'
export const DECREASE_ITEM_TO_CART = 'DECREASE_ITEM_TO_CART'
export const REMOVE_ITEM_TO_CART = 'REMOVE_ITEM_TO_CART'

export const SEARCHCATS = 'SEARCHCATS'

export const FETCH_CAT = 'FETCH_CAT'
export const FETCH_CAT_FAIL = 'FETCH_CAT_FAIL'
export const FETCH_CAT_SUCCESS = 'FETCH_CAT_SUCCESS'

export const RESET_FETCH_CATS = 'RESET_FETCH_CATS'

export const TOGGLE_DRAWER = 'TOGGLE_DRAWER'

export type DrawerType = 'left' | 'right' | 'top' | 'bottom'
export type DrawerPayload = {
  anchor: DrawerType
  open: boolean
}
export type ToggleDrawerAction = {
  type: typeof TOGGLE_DRAWER
  payload: DrawerPayload
}
export type DrawerState = {
  [c in DrawerType]: boolean
}
export type FetchStatusState = {
  fetchCats: boolean | null
}
export type FetchStatusAction =
  | FetchCatFailAction
  | SagaFetchCatSuccessAction
  | ResetFetchCatStatusAction

export type AddItemToCartAction = {
  type: typeof ADD_ITEM_TO_CART
  payload: CatType
}

export type DecreaseItemToCartAction = {
  type: typeof DECREASE_ITEM_TO_CART
  payload: CatType
}

export type RemoveItemToCartAction = {
  type: typeof REMOVE_ITEM_TO_CART
  payload: string
}

export type ResetFetchCatStatusAction = {
  type: typeof RESET_FETCH_CATS
}

export type FetchCatAction = {
  type: typeof FETCH_CAT | typeof FETCH_CAT_FAIL
  payload: CatType[]
}

export type SagaFetchCatAction = {
  type: typeof FETCH_CAT
  payload: string | undefined
}

export type SagaFetchCatSuccessAction = {
  type: typeof FETCH_CAT_SUCCESS
  payload: CatType[]
}

export type FetchCatFailAction = {
  type: typeof FETCH_CAT_FAIL
}

export type CartActions =
  | AddItemToCartAction
  | RemoveItemToCartAction
  | DecreaseItemToCartAction

export type CatActions = FetchCatAction | FetchCatFailAction

export type SagaCatActions =
  | FetchCatFailAction
  | SagaFetchCatSuccessAction

export type CartState = { cats: { [code: string]: CatType } }

export type CatState = { [id: string]: CatType }

// Enum
export enum DialogType {
  SignIn = 'signIn',
  SignUp = 'signUp',
}

// A product
export type Product = {
  id: string
  name: string
  price: number
}

export type AddProductAction = {
  type: typeof ADD_PRODUCT
  payload: {
    product: Product
  }
}

export type RemoveProductAction = {
  type: typeof REMOVE_PRODUCT
  payload: {
    product: Product
  }
}

export type ToggleDialogAction = {
  type: typeof TOGGLE_DIALOG
  payload: {
    dialog: DialogType
  }
}

export type UiActions = ToggleDialogAction | ToggleDrawerAction

// Use this union in reducer
export type ProductActions = AddProductAction | RemoveProductAction

export type ProductState = {
  inCart: Product[]
}

// Using dynamic keys from an enum
export type UiState = {
  dialogOpen: {
    [key in DialogType]?: boolean
  }
  drawer: DrawerState
}

export type AppState = {
  product: ProductState
  ui: UiState
  cart: CartState
  cat: CatState
  fetchStatus: FetchStatusState
  debug: boolean
}

export type CatType = { orderAmount: number } & {
  adaptability: number,
  affection_level: number, 
  alt_names: string,
  cfa_url: string,
  child_friendly: number,
  description: string,
  dog_friendly: number
  energy_level: number,
  experimental: number,
  grooming: number,
  hairless: number,
  health_issues: number,
  hypoallergenic: number,
  id: string,
  image: any,
  indoor: number
  intelligence: number,
  lap: number,
  life_span: string,
  name: string,
  natural: string,
  origin: string,
  rare: number,
  reference_image_id: string,
  rex: number,
  shedding_level: number,
  short_legs: number,
  social_needs: number,
  stranger_friendly: number,
  suppressed_tail: number,
  temperament: string,
  vcahospitals_url: string,
  vetstreet_url: string,
  vocalisation: string,
  weight: any
  wikipedia_url: string
}
export type ErrorType = {
  response: {
    message?: string
    status: number
  }
}

export type ThemeLabel = 'red' | 'blue' | 'green'
export type MainPalette = {
  primary: string
  secondary: string
  disable: string
}
export type SubPalette = {
  primary_contrast: string
  secondary_contrast: string
  disable_contrast: string
}
export type Palette = MainPalette & SubPalette
export type AppTheme = { [c in ThemeLabel]: Palette }

export type ThemeContextType = {
  theme: ThemeLabel
  setTheme: (c: ThemeLabel) => void
}
