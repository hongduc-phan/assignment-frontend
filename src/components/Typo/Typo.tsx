import { ReactNode, createElement, AllHTMLAttributes, Component } from 'react'
import clsx from 'clsx'
import { TypoColors, TypoVariants, TypoAlignment } from './types'

import styles from './Typo.module.css'

export interface TypoProps<T> extends AllHTMLAttributes<T> {
  children?: ReactNode
  align?: TypoAlignment
  className?: string
  color?: TypoColors
  variant?: TypoVariants
  tag?: string
}

class Typo<T = HTMLSpanElement> extends Component<TypoProps<T>> {
  private static getColor(color: string) {
    switch (color) {
    case TypoColors.inherit:
      return styles.inherit
    case TypoColors.black:
      return styles.black
    case TypoColors.white:
      return styles.white
    case TypoColors.greyDark:
      return styles.greyDark
    case TypoColors.greyMedium:
      return styles.greyMedium
    case TypoColors.blue:
      return styles.blue
    case TypoColors.red:
      return styles.red
    default:
      return styles.black
    }
  }

  private static getVariant(variant: string) {
    switch (variant) {
    case TypoVariants.h1:
      return styles.h1
    case TypoVariants.h2:
      return styles.h2
    case TypoVariants.h3:
      return styles.h3
    case TypoVariants.h4:
      return styles.h4
    case TypoVariants.h5:
      return styles.h5
    case TypoVariants.body1:
      return styles.body1
    case TypoVariants.body2:
      return styles.body2
    case TypoVariants.button:
      return styles.button
    case TypoVariants.subTitle:
      return styles.subTitle
    default:
      return styles.body1
    }
  }

  private static getAlign(align: string) {
    if (TypoAlignment.middle === align) {
      return styles.middle
    }

    return styles.left
  }

  public render() {
    const {
      children,
      align = 'left',
      color = TypoColors.black,
      variant = TypoVariants.body1,
      tag = 'span',
      className,
      ...other
    } = this.props
    return createElement(
      tag,
      {
        ...other,
        className: clsx(
          styles.typo,
          className,
          Typo.getVariant(variant),
          Typo.getColor(color),
          Typo.getAlign(align)
        ),
      },
      children
    )
  }
}

export default Typo
