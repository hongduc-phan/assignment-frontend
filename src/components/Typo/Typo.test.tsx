import React from 'react'
import renderer, { ReactTestRendererJSON } from 'react-test-renderer'
import { mount, ReactWrapper, configure }  from 'enzyme'

import Adapter from 'enzyme-adapter-react-16'

import Typo from './Typo'
import { TypoColors, TypoAlignment, TypoVariants } from './types'

configure({ adapter: new Adapter() })

describe('<Typo />', () => {
  let typo: any
  let tree: ReactTestRendererJSON | null
  let component: ReactWrapper

  it('Should display a <span> tag', () => {
    typo = <Typo>Lorem</Typo>
    tree = renderer.create(typo).toJSON() as ReactTestRendererJSON
    expect(tree).toMatchSnapshot()
  })

  it('Should display a <div> tag', () => {
    typo = <Typo tag="div">Lorem</Typo>
    tree = renderer.create(typo).toJSON() as ReactTestRendererJSON
    expect(tree).toMatchSnapshot()
  })

  describe('Colors', () => {
    it('Should display with inherit color', () => {
      typo = <Typo color={TypoColors.inherit}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.inherit')
      expect(spans.length).toEqual(1)
    })

    it('Should display with black color', () => {
      typo = <Typo color={TypoColors.black}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.black')
      expect(spans.length).toEqual(1)
    })

    it('Should display with white color', () => {
      typo = <Typo color={TypoColors.white}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.white')
      expect(spans.length).toEqual(1)
    })

    it('Should display with greyDark color', () => {
      typo = <Typo color={TypoColors.greyDark}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.greyDark')
      expect(spans.length).toEqual(1)
    })

    it('Should display with greyMedium color', () => {
      typo = <Typo color={TypoColors.greyMedium}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.greyMedium')
      expect(spans.length).toEqual(1)
    })

    it('Should display with blue color', () => {
      typo = <Typo color={TypoColors.blue}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.blue')
      expect(spans.length).toEqual(1)
    })

    it('Should display with red color', () => {
      typo = <Typo color={TypoColors.red}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.red')
      expect(spans.length).toEqual(1)
    })
  })

  describe('Alignment', () => {
    it('Should display with align middle', () => {
      typo = <Typo align={TypoAlignment.middle}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.middle')
      expect(spans.length).toEqual(1)
    })

    it('Should display with align left', () => {
      typo = <Typo align={TypoAlignment.left}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.left')
      expect(spans.length).toEqual(1)
    })
  })

  describe('Variants', () => {
    it('Should display with h1', () => {
      typo = <Typo variant={TypoVariants.h1}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.h1')
      expect(spans.length).toEqual(1)
    })

    it('Should display with h2', () => {
      typo = <Typo variant={TypoVariants.h2}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.h2')
      expect(spans.length).toEqual(1)
    })

    it('Should display with h3', () => {
      typo = <Typo variant={TypoVariants.h3}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.h3')
      expect(spans.length).toEqual(1)
    })

    it('Should display with h4', () => {
      typo = <Typo variant={TypoVariants.h4}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.h4')
      expect(spans.length).toEqual(1)
    })

    it('Should display with h5', () => {
      typo = <Typo variant={TypoVariants.h5}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.h5')
      expect(spans.length).toEqual(1)
    })

    it('Should display with body1', () => {
      typo = <Typo variant={TypoVariants.body1}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.body1')
      expect(spans.length).toEqual(1)
    })

    it('Should display with body2', () => {
      typo = <Typo variant={TypoVariants.body2}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.body2')
      expect(spans.length).toEqual(1)
    })

    it('Should display with button', () => {
      typo = <Typo variant={TypoVariants.button}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.button')
      expect(spans.length).toEqual(1)
    })

    it('Should display with subTitle', () => {
      typo = <Typo variant={TypoVariants.subTitle}>Lorem</Typo>
      component = mount(typo)
      const spans = component.find('.subTitle')
      expect(spans.length).toEqual(1)
    })
  })
})