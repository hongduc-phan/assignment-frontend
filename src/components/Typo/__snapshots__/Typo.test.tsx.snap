// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typo /> Should display a <div> tag 1`] = `
<div
  className="typo body1 black left"
>
  Lorem
</div>
`;

exports[`<Typo /> Should display a <span> tag 1`] = `
<span
  className="typo body1 black left"
>
  Lorem
</span>
`;
